src/cata_tiles.cpp

```diff
@@ -179,66 +215,221 @@ void cata_tiles::get_tile_information(std::string config_path, std::string &json
     }
 }
 
+inline static pixel get_pixel_color(SDL_Surface *surf, int x, int y, int w)
+{
+    pixel pix;
+    const auto pixelarray = reinterpret_cast<unsigned char *>(surf->pixels);
+    const auto pixel_ptr = pixelarray + (y * w + x) * 4;
+    pix.r = pixel_ptr[0];
+    pix.g = pixel_ptr[1];
+    pix.b = pixel_ptr[2];
+    pix.a = pixel_ptr[3];
+    return pix;
+}
+
+inline static void set_pixel_color(SDL_Surface *surf, int x, int y, int w, pixel pix)
+{
+    const auto pixelarray = reinterpret_cast<unsigned char *>(surf->pixels);
+    const auto pixel_ptr = pixelarray + (y * w + x) * 4;
+    pixel_ptr[0] = static_cast<unsigned char>(pix.r);
+    pixel_ptr[1] = static_cast<unsigned char>(pix.g);
+    pixel_ptr[2] = static_cast<unsigned char>(pix.b);
+    pixel_ptr[3] = static_cast<unsigned char>(pix.a);
+}
+
+static void color_pixel_grayscale(pixel& pix)
+{
+    bool isBlack = pix.isBlack();
+    int result = (pix.r + pix.b + pix.g) / 3;
+    result = result * 6 / 10;
+    if (result > 255) {
+        result = 255;
+    }
+    //workaround for color key 0 on some tilesets
+    if (result<1 && !isBlack){
+        result = 1;
+    }
+    pix.r = result;
+    pix.g = result;
+    pix.b = result;
+}
+
+static void color_pixel_nightvision(pixel& pix)
+{
+    int result = (pix.r + pix.b + pix.g) / 3;
+    result = result * 3 / 4 + 64;
+    if (result > 255) {
+        result = 255;
+    }
+    pix.r = result / 4;
+    pix.g = result;
+    pix.b = result / 7;
+}
+
+static void color_pixel_overexposed(pixel& pix)
+{
+    int result = (pix.r + pix.b + pix.g) / 3;
+    result = result / 4 + 192;
+    if (result > 255) {
+        result = 255;
+    }
+    pix.r = result / 4;
+    pix.g = result;
+    pix.b = result / 7;
+}
+
+void cata_tiles::apply_color_filter(SDL_Surface *surf, COLOR_FILTER filter)
+{
+    for (int y = 0; y < surf->h; y++) {
+        for (int x = 0; x < surf->w; x++) {
+            pixel pix = get_pixel_color(surf, x, y, surf->w);
+            switch (filter){
```

Comment:
Technically, you don't need a `COLOR_FILTER` enum at all, you can just forward a reference to the function that should be applied to each pixel. And `apply_color_filter` can be a free function.

```C++
void apply_color_filter( SDL_Surface *surf, void (&pixel_converter)(pixel &) )
{
    for( ... ) {
        pixel pix = get_pixel_color( ... );
        pixel_converter( pix );
        set_pixel_color( ... );
    }
}

void init() {
    ...
    apply_color_filter( shadow_tile_atlas, color_pixel_grayscale );
    apply_color_filter( shadow_tile_atlas, color_pixel_overexposed );
    apply_color_filter( shadow_tile_atlas, color_pixel_nightvision );
    ...
}
```

If you keep the enum, please remove the `default:` case in the switch. It disables a very useful warning of the compiler: if the switch statement is over an enum type (as in this case), the compiler will warn you if some entries of the enum are not handled. This comes in handy when another entry is added to the enum, but the switch is not updated.